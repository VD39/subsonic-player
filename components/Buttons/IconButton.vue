<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: string;
    iconPosition?: IconPosition;
    iconSize?: number;
    is?: ButtonType;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconPosition: 'left',
    iconSize: 20,
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      $style.iconButton,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        [$style.fullWidth]: fullWidth,
      },
    ]"
    :disabled="disabled"
    :type="type"
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="phIcon"
      :class="$style.icon"
      aria-hidden="true"
      :size="iconSize"
      weight="duotone"
    />

    <span
      ref="text"
      :class="[
        $style.text,
        {
          'visually-hidden': !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.iconButton {
  @mixin align-center;

  position: relative;
  gap: var(--space-8);
  padding: var(--space-8);
  margin: 0;
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.fullWidth {
  width: 100%;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  fill: var(--body-font-color);

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
</style>
