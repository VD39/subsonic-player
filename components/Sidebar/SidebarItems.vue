<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { collapsed } = useSidebar();

withDefaults(
  defineProps<{
    items?: Navigation[];
    title?: string;
    titleIcon?: string;
    titleTo?: string;
  }>(),
  {
    items: () => [],
    title: undefined,
    titleIcon: undefined,
    titleTo: undefined,
  },
);
</script>

<template>
  <li :class="$style.sidebarItem">
    <NuxtLink
      v-if="title"
      ref="rootTitle"
      :to="titleTo"
      :class="$style.titleLink"
      :title="title"
    >
      <component
        :is="titleIcon"
        v-if="titleIcon"
        :class="$style.titleIcon"
        aria-hidden="true"
        weight="duotone"
        :size="32"
      />

      <h4 :class="$style.title">
        {{ title }}
      </h4>
    </NuxtLink>

    <ul v-if="items.length" ref="items" :class="$style.items">
      <li v-for="(item, index) in items" :key="`item-${index}`">
        <IconButton
          is="nuxt-link"
          v-if="!item.component"
          :to="item.to"
          :icon="item.icon"
          :class="$style.link"
          :exact-active-class="$style.current"
          :show-text="!collapsed"
          :title="item.title"
        >
          {{ item.title }}
        </IconButton>

        <component
          :is="item.component"
          v-if="item.component"
          data-test-id="item-component"
        />
      </li>
    </ul>
  </li>
</template>

<style module>
.sidebarItem {
  @media (--tablet-up) {
    margin-bottom: var(--space-16);

    :global(.collapsed) & {
      margin-bottom: 0;
    }
  }
}

.titleLink {
  @mixin align-center;

  flex-direction: column;
  justify-content: center;
  margin-bottom: 0;

  @media (--tablet-up) {
    display: unset;
    pointer-events: none;
  }
}

.titleIcon {
  @media (--tablet-up) {
    display: none;
  }
}

.title {
  @media (--mobile-only) {
    margin-bottom: 0;
    font-size: var(--small-font-size);
  }

  @media (--tablet-up) {
    padding-left: var(--space-24);

    :global(.collapsed) & {
      @mixin visually-hidden;
    }
  }
}

.items {
  display: none;

  @media (--tablet-up) {
    display: block;
  }
}

.link {
  @mixin sidebar-link;
}

.current {
  @mixin sidebar-current-link;
}
</style>
