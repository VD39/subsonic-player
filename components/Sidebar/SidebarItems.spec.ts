import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import IconButton from '@/components/Buttons/IconButton.vue';
import SidebarItems from './SidebarItems.vue';

const collapsedMock = ref(false);

mockNuxtImport('useSidebar', () => () => ({
  collapsed: collapsedMock,
}));

function factory(props = {}) {
  return mount(SidebarItems, {
    props: {
      ...props,
    },
  });
}

describe('SidebarItems', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when title prop is not set', () => {
    it('does not show the title', () => {
      expect(wrapper.findComponent({ ref: 'rootTitle' }).exists()).toBe(false);
    });
  });

  describe('when title prop is set to true', () => {
    beforeEach(() => {
      wrapper = factory({
        title: 'title',
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the title', () => {
      expect(wrapper.findComponent({ ref: 'rootTitle' }).exists()).toBe(true);
    });
  });

  describe('when items prop is not set', () => {
    it('does not show the title', () => {
      expect(wrapper.find({ ref: 'items' }).exists()).toBe(false);
    });
  });

  describe('when items prop is set', () => {
    beforeEach(() => {
      wrapper = factory({
        items: [
          {
            name: 'name',
            title: 'title',
            to: 'to',
            icon: 'icon',
          },
          {
            name: 'name',
            title: 'title',
            to: 'to',
            icon: 'icon',
          },
          {
            component: 'div',
          },
        ],
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the items', () => {
      expect(wrapper.find({ ref: 'items' }).exists()).toBe(true);
    });

    it('shows the correct amount of item', () => {
      expect(wrapper.findAllComponents(IconButton).length).toBe(2);
    });

    it('shows the correct amount of components', () => {
      expect(wrapper.findAll('[data-test-id="item-component"]').length).toBe(1);
    });

    describe('when collapsed value is false', () => {
      it('sets the correct show-text prop on the IconButton component', () => {
        expect(wrapper.findAllComponents(IconButton)[0].props('showText')).toBe(
          true,
        );
      });
    });

    describe('when collapsed value is true', () => {
      beforeEach(() => {
        collapsedMock.value = true;
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('sets the correct show-text prop on the IconButton component', () => {
        expect(wrapper.findAllComponents(IconButton)[0].props('showText')).toBe(
          false,
        );
      });
    });
  });
});
