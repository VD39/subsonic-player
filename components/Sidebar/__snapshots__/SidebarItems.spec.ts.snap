// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarItems > matches the snapshot 1`] = `
"<li class="sidebarItem">
  <!--v-if-->
  <!--v-if-->
</li>"
`;

exports[`SidebarItems > when items prop is set > matches the snapshot 1`] = `
"<li class="sidebarItem">
  <!--v-if-->
  <ul class="items">
    <li><a exactactiveclass="current" class="iconButton link link" title="title">
        <icon class="icon" aria-hidden="true" size="20" weight="duotone"></icon><span class="text">title</span>
      </a>
      <!--v-if-->
    </li>
    <li><a exactactiveclass="current" class="iconButton link link" title="title">
        <icon class="icon" aria-hidden="true" size="20" weight="duotone"></icon><span class="text">title</span>
      </a>
      <!--v-if-->
    </li>
    <li>
      <!--v-if-->
      <div data-test-id="item-component"></div>
    </li>
  </ul>
</li>"
`;

exports[`SidebarItems > when items prop is set > when collapsed value is true > matches the snapshot 1`] = `
"<li class="sidebarItem">
  <!--v-if-->
  <ul class="items">
    <li><a exactactiveclass="current" class="iconButton link link" title="title">
        <icon class="icon" aria-hidden="true" size="20" weight="duotone"></icon><span class="text visually-hidden">title</span>
      </a>
      <!--v-if-->
    </li>
    <li><a exactactiveclass="current" class="iconButton link link" title="title">
        <icon class="icon" aria-hidden="true" size="20" weight="duotone"></icon><span class="text visually-hidden">title</span>
      </a>
      <!--v-if-->
    </li>
    <li>
      <!--v-if-->
      <div data-test-id="item-component"></div>
    </li>
  </ul>
</li>"
`;

exports[`SidebarItems > when title prop is set to true > matches the snapshot 1`] = `
"<li class="sidebarItem"><a class="titleLink" title="title">
    <!--v-if-->
    <h4 class="title">title</h4>
  </a>
  <!--v-if-->
</li>"
`;
